pyomo/core/base/constraint.py

```diff
@@ -614,21 +622,30 @@ def set_value(self, expr):
             raise DeveloperError("Unrecognized relational expression type: %s"
                                  % (self._expr.__class__.__name__,))
 
+        # We have historically forced the body to be a numeric expression.
+        # TODO: remove this requirement
+        if self._body.__class__ in native_types and self._body is not None:
+            self._body = as_numeric(self._body)
+
         # We have historically mapped incoming inf to None
         if self._lower.__class__ in native_numeric_types:
-            if self._lower == -_inf:
-                self._lower = None
-            elif not math.isfinite(self._lower):
-                raise ValueError(
-                    "Constraint '%s' created with an invalid non-finite "
-                    "lower bound (%s)." % (self.name, self._lower))
+            bound = self._lower
+            if bound in _nonfinite_values or bound != bound:
+                if bound == -_inf:
+                    self._lower = None
+                else:
+                    raise ValueError(
+                        "Constraint '%s' created with an invalid non-finite "
+                        "lower bound (%s)." % (self.name, self._lower))
         if self._upper.__class__ in native_numeric_types:
-            if self._upper == _inf:
-                self._upper = None
-            elif not math.isfinite(self._upper):
-                raise ValueError(
-                    "Constraint '%s' created with an invalid non-finite "
-                    "upper bound (%s)." % (self.name, self._upper))
+            bound = self._upper
+            if bound in _nonfinite_values or bound != bound:
```

Comment:
Same question as above.

Reply:
If bound is `float('nan')`, then `(bound != bound) is True`